oadp-operator/controllers/registry.go

```diff
@@ -0,0 +1,356 @@
+func (r *VeleroReconciler) buildRegistryDeployment(registryDeployment *appsv1.Deployment, bsl *velerov1.BackupStorageLocation) *appsv1.Deployment {
+	registryDeployment.Labels = r.getBSLLabels(bsl)
+
+	registryDeployment.Spec = appsv1.DeploymentSpec{
+		Replicas: pointer.Int32(1),
+		Template: corev1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: map[string]string{
+					"component": constructBSLRegistryName(bsl),
+				},
+			},
+			Spec: corev1.PodSpec{
+				RestartPolicy: corev1.RestartPolicyAlways,
+				Containers:    r.buildRegistryContainer(bsl),
+			},
+		},
+	}
+	return registryDeployment
+}
+
+func (r *VeleroReconciler) getBSLLabels(bsl *velerov1.BackupStorageLocation) map[string]string {
+	labels := map[string]string{
+		"app.kubernetes.io/name":       BackupStorageLocation,
+		"app.kubernetes.io/instance":   constructBSLRegistryName(bsl),
+		"app.kubernetes.io/managed-by": OADPOperator,
+		"app.kubernetes.io/component":  Registry,
+	}
+	return labels
+}
+
+func constructBSLRegistryName(bsl *velerov1.BackupStorageLocation) string {
+	return "oadp-" + bsl.Name + "-" + bsl.Spec.Provider + "-registry"
+}
+
+func (r *VeleroReconciler) buildRegistryContainer(bsl *velerov1.BackupStorageLocation) []corev1.Container {
+	containers := []corev1.Container{
+		{
+			Image: RegistryImage,
+			Name:  constructBSLRegistryName(bsl) + "-container",
+			Ports: []corev1.ContainerPort{
+				{
+					ContainerPort: 5000,
+					Protocol:      corev1.ProtocolTCP,
+				},
+			},
+			Env: getRegistryEnvVars(bsl),
+			LivenessProbe: &corev1.Probe{
+				Handler: corev1.Handler{
+					HTTPGet: &corev1.HTTPGetAction{
+						Path: "/v2/_catalog?n=5",
+						Port: intstr.IntOrString{IntVal: 5000},
+					},
+				},
+				PeriodSeconds:       5,
+				TimeoutSeconds:      3,
+				InitialDelaySeconds: 15,
+			},
+			ReadinessProbe: &corev1.Probe{
+				Handler: corev1.Handler{
+					HTTPGet: &corev1.HTTPGetAction{
+						Path: "/v2/_catalog?n=5",
+						Port: intstr.IntOrString{IntVal: 5000},
+					},
+				},
+				PeriodSeconds:       5,
+				TimeoutSeconds:      3,
+				InitialDelaySeconds: 15,
+			},
+		},
+	}
+
+	return containers
+}
+
+func getRegistryEnvVars(bsl *velerov1.BackupStorageLocation) []corev1.EnvVar {
+	envVar := []corev1.EnvVar{}
+	if bsl.Spec.Provider == AWSProvider {
+		envVar = buildRegistryEnvVars(bsl, cloudProviderEnvVarMap[AWSProvider], AWSProvider)
+	}
+
+	if bsl.Spec.Provider == AzureProvider {
+		envVar = buildRegistryEnvVars(bsl, cloudProviderEnvVarMap[AzureProvider], AzureProvider)
+	}
+
+	if bsl.Spec.Provider == GCPProvider {
+		envVar = buildRegistryEnvVars(bsl, cloudProviderEnvVarMap[GCPProvider], GCPProvider)
+	}
+
+	return envVar
+}
+
+func buildRegistryEnvVars(bsl *velerov1.BackupStorageLocation, providerEnvVar []corev1.EnvVar, provider string) []corev1.EnvVar {
+
+	updatedEnvVar := providerEnvVar
+	// build AWS provider env vars
+	if provider == AWSProvider {
+		for _, key := range updatedEnvVar {
```

Comment:
I don't think this will work. You are not modifying the original array, you are modifying a copy of the element in the array. I would rework this to instead have functions like `getAwsEnvVars()` that returns `[]corev1.EnvVar` and then return it here.

You can see an example of this not working here: https://play.golang.org/p/HyEisqurot3

Reply:
Thanks for the catch @dymurray Updated the PR.